File: 0123-best-time-to-buy-and-sell-stock-iii/0123-best-time-to-buy-and-sell-stock-iii.java
class Solution {
    public int maxProfit(int[] prices) {
        if(prices == null || prices.length < 2) {
            return 0;
        }

        int[] leftProfit = new int[prices.length];
        int[] rightProfit = new int[prices.length];

        int minPrice = Integer.MAX_VALUE;
        int maxProfit = Integer.MIN_VALUE;

        for(int i=0; i<prices.length; i++) {
            minPrice = Math.min(minPrice, prices[i]);
            maxProfit = Math.max(maxProfit, prices[i] - minPrice);
            leftProfit[i] = maxProfit;
        }

        int maxPrice = Integer.MIN_VALUE;
        maxProfit = Integer.MIN_VALUE;

        for(int i=prices.length-1; i>=0; i--) {
            maxPrice = Math.max(maxPrice, prices[i]);
            maxProfit = Math.max(maxProfit, maxPrice - prices[i]);
            rightProfit[i] = maxProfit;
        }

        int max = 0;

        for(int i=0; i<prices.length; i++) {
            max = Math.max(max, leftProfit[i] + rightProfit[i]);
        }

        return max;
    }
}

File: 0123-best-time-to-buy-and-sell-stock-iii/0123-best-time-to-buy-and-sell-stock-iii.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if not prices:
            return 0
    
        hold1 = float('-inf')
        sell1 = 0
        hold2 = float('-inf')
        sell2 = 0
        
        for price in prices:
            hold1 = max(hold1, -price)
            sell1 = max(sell1, hold1 + price)
            hold2 = max(hold2, sell1 - price)
            sell2 = max(sell2, hold2 + price)
            
        return sell2